### PR TITLE
fix: fix vaddr INTR_TVAL on XS_ARCH

### PR DESCRIPTION
**Pull request by xiaokamikami**
like https://github.com/OpenXiangShan/NEMU/pull/284
Refine the changes to support  on  address misaligned，address fault，load and store page fault

### CODE REVIEW

File: src/isa/riscv64/system/intr.c [hunk 1]
```diff
@@ -238,4 +238,11 @@ word_t isa_query_intr() {
     }
   }
   return INTR_EMPTY;
-}
\ No newline at end of file
+}
+
+#ifdef CONFIG_USE_XS_ARCH_CSRS
+word_t INTR_TVAL_SV39_SEXT(word_t vaddr) {
+    vaddr = vaddr & (vaddr_t)0x7FFFFFFFFF;
```
Comment (poemonsense) [inline src/isa/riscv64/system/intr.c:245]: use 2 spaces instead of 4 spaces for indentation

File: src/isa/riscv64/system/mmu.c [hunk 2]
```diff
@@ -530,11 +541,16 @@ int isa_mmu_check(vaddr_t vaddr, int len, int type) {
...
       }else{
         ex = cpu.amo ? EX_SPF : EX_LPF;
+        IFDEF(CONFIG_USE_XS_ARCH_CSRS, vaddr = INTR_TVAL_SV39_SEXT(vaddr));
         INTR_TVAL_REG(ex) = vaddr;
       }
       longjmp_exception(ex);
 #else
       int ex = cpu.amo ? EX_SPF : EX_LPF;
+#ifdef CONFIG_USE_XS_ARCH_CSRS
+      vaddr = INTR_TVAL_SV39_SEXT(vaddr);
```
Comment (poemonsense) [inline src/isa/riscv64/system/mmu.c:551]: duplicated code

File: src/isa/riscv64/system/mmu.c [hunk 3]
```diff
@@ -557,10 +573,18 @@ int isa_mmu_check(vaddr_t vaddr, int len, int type) {
         }
         longjmp_exception(EX_SPF);
       }else{
+#ifdef CONFIG_USE_XS_ARCH_CSRS
+        vaddr = INTR_TVAL_SV39_SEXT(vaddr);
+#endif // CONFIG_USE_XS_ARCH_CSRS
+        IFDEF(CONFIG_USE_XS_ARCH_CSRS, vaddr = INTR_TVAL_SV39_SEXT(vaddr));
```
Comment (poemonsense) [inline src/isa/riscv64/system/mmu.c:579]: duplicated

File: src/isa/riscv64/system/mmu.c [hunk 4]
```diff
@@ -557,10 +573,18 @@ int isa_mmu_check(vaddr_t vaddr, int len, int type) {
...
         INTR_TVAL_REG(EX_SPF) = vaddr;
         longjmp_exception(EX_SPF);
       }
 #else
+#ifdef CONFIG_USE_XS_ARCH_CSRS
+      vaddr = INTR_TVAL_SV39_SEXT(vaddr);
```
Comment (poemonsense) [inline src/isa/riscv64/system/mmu.c:585]: duplicated

File: src/isa/riscv64/system/mmu.c [hunk 5]
```diff
@@ -583,11 +607,10 @@ void isa_misalign_data_addr_check(vaddr_t vaddr, int len, int type) {
     Log("addr misaligned happened: vaddr:%lx len:%d type:%d pc:%lx", vaddr, len, type, cpu.pc);
     int ex = cpu.amo || type == MEM_TYPE_WRITE ? EX_SAM : EX_LAM;
 #ifdef CONFIG_USE_XS_ARCH_CSRS
-    vaddr = vaddr & (vaddr_t)0x7FFFFFFFFF;
-    INTR_TVAL_REG(ex) = SEXT(vaddr, 39); // USE SV39 VADDR
-#else
-    INTR_TVAL_REG(ex) = vaddr;
+    vaddr = INTR_TVAL_SV39_SEXT(vaddr);
```
Comment (poemonsense) [inline src/isa/riscv64/system/mmu.c:610]: also duplicated?

**Review by poemonsense — COMMENTED** [review]
Looks OK to me. But minor code changes are required. Some code seems to be duplicated

File: src/isa/riscv64/system/intr.c (hunk 1 above)
Comment (xiaokamikami) [reply src/isa/riscv64/system/intr.c]: ok

**Pull request merged**